Alongside OS & Refresh Install Options

mmstick opened the pull request:
Requires https://github.com/pop-os/distinst/pull/132

## Refresh Install

This takes Cassidy's work with the DecryptionDialog at https://github.com/elementary/installer/pull/333, and integrates it with distinst, as well as a new RefreshView to list detected refresh options and to make them happen. This will require a few changes in distinst to expose some previously-hidden APIs, and to fix some issues with installing (updating the recovery partition fails due to running out of space).

### Screenshots

![screenshot from 2018-11-08 11-30-22](https://user-images.githubusercontent.com/4143535/48219585-1ee1ce00-e34a-11e8-9582-2a531adc5b18.png)

![screenshot from 2018-11-08 16-31-08](https://user-images.githubusercontent.com/4143535/48233751-b90a3c00-e373-11e8-97d8-87b154a28655.png)



### Testing

Perform a clean install with and without encryption, and then do the following:

- Create a few user accounts, with some files
- Reboot and use the refresh install option the second time around
- Validate that the files are there, and that the user is still part of the sudo group

For the UI, these can be tested:

- Typing the wrong password / duplicate PV in the decryption menu should display an error dialog
- When only one refresh option is available, the UI should immediately progress to installing it
- When multiple options are possible, there's a UI that will let the user select which one to choose.
- If no LUKS devices are discovered, then the decryption dialog bar on the "Try Install" screen should be hidden by default.
- After decrypting all devices that can be decrypted, the dialog bar should be hidden.
- If multiple devices were found, and a device has been decrypted, an unlock icon should be on that partition's entry in the decryption dialog, instead of a locked icon.

## Alongside OS Install

The work had already been done in distinst, so it was just a matter of exposing it in the UI. By sharing UI code with the Refresh install view, both views have the same option selection view, but the alongside choice will also feature a slider for determining how much a partition should be shrunk to make room for the new install.

### Screenshots

![screenshot from 2018-11-08 17-09-50](https://user-images.githubusercontent.com/4143535/48234983-22407e00-e379-11e8-8f8e-e97c7555c37b.png)

#### Sector Slider


### Testing

- Perform an unencrypted install, and then attempting to perform an unencrypted and/or encrypted install next to it.
- Attempt to install alongside Windows
- Make a large region of free space, next to an install, and install to that.

jackpot51 commented:
@mmstick Please fix the build issues

mmstick commented:
@jackpot51 It should build once https://github.com/pop-os/distinst/pull/132 is merged, since it requires some FFI and script changes.

WatchMkr commented:
Why would a user want to keep the backup install? What would they do with it?

mmstick commented:
New screenshots

![screenshot from 2018-09-05 14-02-41](https://user-images.githubusercontent.com/4143535/45118649-4af88d00-b116-11e8-9a0d-ff74eb7efe86.png)
![screenshot from 2018-09-05 14-02-56](https://user-images.githubusercontent.com/4143535/45118684-62d01100-b116-11e8-890a-6be14421295e.png)
![screenshot from 2018-09-05 14-03-08](https://user-images.githubusercontent.com/4143535/45118661-51870480-b116-11e8-9e0b-bdd5f79e238d.png)
![screenshot from 2018-09-05 14-03-19](https://user-images.githubusercontent.com/4143535/45118668-58157c00-b116-11e8-9c93-31193d11a01b.png)
![screenshot from 2018-09-05 14-14-19](https://user-images.githubusercontent.com/4143535/45118673-5b106c80-b116-11e8-9e33-12ba469367ac.png)

WatchMkr commented:
Few things:

Let's move the "Applications installed on the system...." copy to the previous page "Refresh Install" description. To shorten it up we can use "Reinstall while keeping user accounts and files. Applications will need to be reinstalled manually." This way we can remove the extra step if there's only one possible target.

Since the drive is unlocked in the third screenshot, should the unlock dialog be removed?

Regarding prompting to keep the old install, I don't like this option for a few reasons.
-- As a user, the affirmative action is to choose the green button; however, I have no utility to restore to this install. I don't know why I should choose yes or no. My instinct is to keep the backup just-in-case but that's likely unnecessary and a poor default choice.
-- It breaks with the application UI (fixable)
-- I don't know how much space it will take. (fixable)
-- We're treating it somewhat like a snapshot backup but not considering other methods and technologies that might do this better. I'd prefer to consider snapshots and backups outside of the installer before thinking through how it affects the installer.

mmstick commented:
@WatchMkr 

> Since the drive is unlocked in the third screenshot, should the unlock dialog be removed?

Yes, I'm working on the means to do that.

I do like the option to keep it, myself, but I'll remove it for now.

mmstick commented:
I think this is ready for some testing. This will require to be tested alongside https://github.com/pop-os/distinst/pull/132, which will need basically every part of the installer to be tested again.

As soon as this is merged, I'll be adding the multiboot install option to the UI. An upgrade option could also be added to the UI somehow, since the installer can now compare the name and version of the OS to install with the name and version of the OS being refreshed.

mmstick commented:
Found an issue with secondary groups. Will have that fixed soon.

mmstick commented:
The "Install Alongside OS" option is ready now. It will list up to two options per block device: installing to the largest free region that meets minimum requiements, and shrinking the partition with the largest unused space.

brs17 commented:
The first thing I noticed with this pr is the need to scroll to see all of the installer options. I think that makes the installer feel heavier and is probably not great UX.

![screenshot from 2018-11-07 19-53-15](https://user-images.githubusercontent.com/7460774/48158066-b2ee5f80-e28e-11e8-9be1-d55954a15df4.png)

brs17 commented:
The first thing I tried to do was do a "Refresh Install". The install I had on the disk was a custom partitioned drive (efi, /swap, /, /home, /tmp). After selecting this option, the installer immediately failed:

![screenshot from 2018-11-07 19-53-48](https://user-images.githubusercontent.com/7460774/48158695-67d54c00-e290-11e8-8d82-1d5045efa4fe.png)

I then selected to try installing again. This time when selecting "Refresh Install", I got the next page.

Until I noticed that the button to proceed was enabled. See the following screenshot:
![screenshot from 2018-11-07 19-54-28](https://user-images.githubusercontent.com/7460774/48158967-3610b500-e291-11e8-9db5-9861b2a4c34d.png)
 
It seems as though something is missing here (as it differs from the screenshot you provided).

Proceeding here, this fails as well
![screenshot from 2018-11-07 19-54-46](https://user-images.githubusercontent.com/7460774/48159045-73754280-e291-11e8-904d-4595e2e34350.png)

brs17 commented:
With the previous having failed, I then attempted to "Install alongside OS". 
![screenshot from 2018-11-07 19-59-17](https://user-images.githubusercontent.com/7460774/48159470-84728380-e292-11e8-8ac0-f9e21130afea.png)

This seems to be quite text heavy and potentially confusing in what is going to happen exactly.

Once selecting one of the options, the selection itself looks a little odd, with the orange highlight so far to the left of the actual option (since the button is that wide).

Also I am not really sure here what partitions those options are. (which one is partition 4?)

![screenshot from 2018-11-07 20-02-53](https://user-images.githubusercontent.com/7460774/48159597-e7fcb100-e292-11e8-985d-7ac4ecb514c3.png)

My confusion continues on the next page:

![screenshot from 2018-11-07 20-03-53](https://user-images.githubusercontent.com/7460774/48159739-49248480-e293-11e8-9819-8558c43cb6dd.png)

I think this could be shown better to explain what is happening to both partitions. 

Moving the slider all the way to the left, I have a non-zero value. Why is that? Is MiB the best unit to use here with >9000MiB being the value I am looking at? With custom partitioning we just use gparted for all of the partitioning work, why not continue that here (and benefit from the visual contexts it provides)?

mmstick commented:
Seems the recent distinst changes have broken the options. Probably the PARTUUID change.

> Moving the slider all the way to the left, I have a non-zero value.

The minimum value for the slider is the minimum required size for an install. There has to be enough room for all the partitions that distinst will create in the free space of that shrunk partition.

> With custom partitioning we just use gparted for all of the partitioning work, why not continue that here (and benefit from the visual contexts it provides)?

Main reason is that they don't need to. The installer will automate that process for them, whether it's choosing to install to unpartitioned space on the disk, or shrinking an existing partition to make room for the new install.

WatchMkr commented:
Try or Install page UX Changes:

Change the page title from "Try or Install" to "Install"
Change "Try Demo Mode" to a button in the ancillary action location at the bottom left of the window.

Copy change:
Install Alongside another Operating System
Add Pop!_OS next to one or more existing OS installations.

Alignment: Please align the Refresh Install option title and description with the other options.

I'll take a peak at the refresh and install alongside OS options next.

brs17 commented:
When selecting to `Unlock Encrypted Storage` a popup appears:

![screenshot from 2018-11-07 22-47-26](https://user-images.githubusercontent.com/7460774/48167301-8c3d2280-e2a8-11e8-963d-19eea5491f75.png)

The only other time a popup appears is when gparted opens. Do we want a popup here? I could follow either argument.

I did notice, one has to press <kbd>ESC</kbd> twice to close the popup which seems odd.

Unlock page:
![screenshot from 2018-11-07 22-48-39](https://user-images.githubusercontent.com/7460774/48167849-51d48500-e2aa-11e8-8d51-8c3aa4a3c20e.png)


Wrong password page:
![screenshot from 2018-11-07 22-48-44](https://user-images.githubusercontent.com/7460774/48167868-6284fb00-e2aa-11e8-9d7d-10acebcd203d.png)

Thoughts @WatchMkr @jackpot51 ?

brs17 commented:
> The minimum value for the slider is the minimum required size for an install. There has to be enough room for all the partitions that distinst will create in the free space of that shrunk partition.
> 
>     With custom partitioning we just use gparted for all of the partitioning work, why not continue that here (and benefit from the visual contexts it provides)?
> 

I guess I kinda felt like a lot of this was presented without as good context as I would have liked.

WatchMkr commented:
I think it's correct to use a pop-up for encryption unlock. If there is only one possible partition to unlock, skip selection and go directly to the second window. The copy for the second window should be:

"Unlock partition /dev/$partname" along with the Password and Device Name entry boxes. Is Device Name already filled in?

brs17 commented:
Device name is already filled in.

I also saw that when typing in an incorrect password, once I pressed back, the incorrect password was still in the field, I imagine that should be cleared as well.

brs17 commented:
I don't know why I didn't think about it sooner but I should probably use Peek or something and just record myself going through the process.

evertiro commented:
> Looking at build: [9aaae3c](https://github.com/pop-os/installer/commit/9aaae3c81519a61d78d359501494199a4cb32743) after doing a full disk encrypted install:
> 
> @WatchMkr, when you suggested moving `Try Demo Mode` to the bottom left, did you mean the following way or making it a button like the `Next` and `Back` button?
> 
> ![screenshot from 2018-11-07 22-41-00](https://user-images.githubusercontent.com/7460774/48166363-96115680-e2a5-11e8-8704-db272cc4c832.png)

Having Install, Refresh, Custom, Alongside, Demo... it gets bulky fast. Is there any good reason we can't do what Ubuntu did and have the first dialog simply be Try/Install, with the installation options being presented only after the user has indicated they want to install?

That would allow us to have a bit of a clear divide between "try" and "install", while still keeping the initial install screen from feeling overly cluttered.

cassidyjames commented:
@evertiro you can check the full design spec at https://github.com/elementary/installer/wiki/Try-or-Install. But essentially the reasoning is that adding another step before we can begin installing just slows the user down. We should be able to offer a few sane options here without it getting too cluttered up.

WatchMkr commented:
@evertiro we'd like to continue the user down the most likely and preferred path. Installing being the primary purpose and using demo mode or the live session being ancillary.

WatchMkr commented:
The Install Alongside feature needs some design direction. There was done already but I can't find it. @cassidyjames do you have the original work on that? I don't see it on the wiki.

evertiro commented:
Oh I 100% agree that things should be streamlined and as simple as possible for users. Unfortunately, as the screenshot posted by @brs17 demonstrates, that particular screen is far from simple... It's downright cluttered! I'm not saying that going with a completely separate window is the _right_ solution, only that there needs to be some sort of division between "try" and "install" for the users who aren't already familiar with Linux.

brs17 commented:
I understand your concern, let's take a look once the "try" button becomes a button like the "next" button and reaccess once we see screenshots

evertiro commented:
> EDIT: I think moving the Try button is what we need to clean it up...note not all the options appear all the time (they are dependent on what is already on the system)

This may be true. I'd like to see what you're envisioning for the Try button though.

cassidyjames commented:
@WatchMkr there wasn't much set in stone. I was actually working on this today in the elementary Slack.  The original (i.e. really old) mockup looked like this:

![mockup](https://lh5.googleusercontent.com/I2EJjOBOVhX0DhIm6BSfTdk5jYGy3Y36iTiecxt-QTxz11iB-jYMNqqWVTuKgDNl4RgeklwvSEaXBoc4uqMjAFPMPkj6ye609gL6-vicgXNmTJujAFZewkyNnbJQeorXkizYGgo)
I'm working on something (VERY in-progress) that looks like this:

![image 10](https://user-images.githubusercontent.com/611168/48170217-b5af7b80-e2b3-11e8-8c80-99e68fed4b63.png)

WatchMkr commented:
That looks great. What if there are more than one OS’s installed?

cassidyjames commented:
@WatchMkr the intent was that that's a different use case; a single other OS is probably common, even for a less advanced user, so we handle that case here. But having multiple OSes likely signals that the user is more comfortable with setting up their installs how they want, so Custom should still be sufficient.

The different scenarios are listed out at https://github.com/elementary/installer/wiki/Try-or-Install

WatchMkr commented:
Sounds good.

mmstick commented:
@WatchMkr That should work regardless of how many operating systems are installed, though the word would change depending on what is being shrunk. Possible install locations could be:

- The largest unused region on a disk, in which case no resizing of an existing OS is required.
- An existing partition that has on OS on it, in which case we are shrinking that OS's partition.
- An existing partition that does not have an OS on it, in which case we're just shrinking a data partition.

mmstick commented:
I've fixed the refresh installs. PR coming soon for distinst.

evertiro commented:
> This is what the install view looks like now, with all the options in place.
> 
> ![screenshot from 2018-11-08 11-30-22](https://user-images.githubusercontent.com/4143535/48219483-dd512300-e349-11e8-8980-5537c99b9bcf.png)

That looks a TON better! Though personally, I think the demo button is a bit understated.

brs17 commented:
@cassidyjames: @WatchMkr, @mmstick, and I just sat down to discuss these new additions to the installer, we like the designs you've put up thus far (and vids we've seen in Slack).

@mmstick and @cassidyjames might be worth being in touch as you two make progress so we aren't reinventing the wheel :stuck_out_tongue:

mmstick commented:
WIP changes to the views

![screenshot from 2018-11-08 16-31-08](https://user-images.githubusercontent.com/4143535/48233751-b90a3c00-e373-11e8-97d8-87b154a28655.png)
![screenshot from 2018-11-08 16-30-50](https://user-images.githubusercontent.com/4143535/48233752-b90a3c00-e373-11e8-9d1d-f87881aea6e8.png)

mmstick commented:
@brs17 This is ready for testing, though you'll have to compile this from source with the `refresh-fixes` branch of distinst installed.

brs17 commented:
@mmstick why does it need to be compiled from source?

mmstick commented:
@brs17 Because this won't build without the `refresh-fixes` version of distinst.

mmstick commented:
Now with OS logos:

![screenshot from 2018-11-14 14-11-51](https://user-images.githubusercontent.com/4143535/48512757-7d5def00-e817-11e8-99b2-285ba101cfe6.png)

Once we get an icon for Pop in our icon theme, that can be displayed as well[0].

[0] https://github.com/pop-os/icon-theme/issues/39

cassidyjames commented on src/Views/ResizeView.vala:
```diff
@@ -51,15 +51,15 @@ public class ResizeView : AbstractInstallerView {
         
         scale.show_fill_level = true;
-        scale.get_style_context ().add_class (Granite.STYLE_CLASS_ACCENT);
+        scale.get_style_context ().add_class ("accent");
```
Considering elementary OS is build from bionic, are you sure? Does Pop!_OS just have an outdated Granite?

mmstick replied:
Builds are failing otherwise, so likely.

mmstick replied:
Pop does not include Granite, so it uses the version in the Ubuntu repository: https://packages.ubuntu.com/bionic/libgranite-dev

cassidyjames replied:
@mmstick ah. That may become a problem down the road with both the installer and AppCenter, so it might be worth copying Granite over from the elementary repo.

mmstick replied:
Importing it now

brs17 commented:
## Testing "Alongside OS" Install
- [ ] [new install + old install]
- [ ] Pop 18.10 + Pop 18.04 encrypted
- [ ] Pop 18.04 + Pop 18.10 encrypted
- [ ] Pop 18.10/18.04 + Ubuntu 18.04(systemdboot + grub)
- [ ] Pop 18.10/18.04 + Windows
- [ ] Pop 18.10/18.04 + Fedora on LVM

---

- [ ] Pop 18.10/18.04 + Fedora 29 + Ubuntu 18.04 + Windows (Windows on separate drive)

---

- [ ] Pop 18.10/18.04 + Pop 18.04/18.10 on different drive
- [ ] Pop 18.10/18.04 + empty partition
- [ ] Pop 18.10/18.04 + empty drive


## Testing "Refresh" Install
- [ ] Test a Refresh install of 18.04/18.10 with an encrypted home folder
- [ ] Pop encrypted
- [ ] Pop unencrypted
- [ ] Refresh Pop in dualboot

brs17 commented:
- "Refresh" option was available when installing two different operating systems (Pop and Ubuntu, Pop 18.10 with 18.04)
- In 18.04 the "refresh" icon is not very visible:
![alongside_1](https://user-images.githubusercontent.com/7460774/49349616-ab8d5c80-f668-11e8-842f-0734255aa612.png)
![alongside_2](https://user-images.githubusercontent.com/7460774/49349620-af20e380-f668-11e8-8d71-4b1d1457f309.png)
- Alongside OS option is not visible when attempting to install Pop!_OS 18.10 alongside an encrypted Pop!_OS 18.04 install (or the other way around). The only option shown was a "Refresh install" which shouldn't be an option in this case. I was able to do an "alongside" install after doing a custom install of the operating system.
- On the "Resize OS" page, the left side of the installer ("Resize OS" text as well as Zoe image) is smaller than the left side content is on other pages.
- When doing an "Alongside" install, The max size to shrink the current install is ~30.3GB (20GB of free space). The minimum size of the new install was 14.3GB (0 bytes free). Why shouldn't both sides require free space?
- When successfully doing an "Alongside OS" install, two efi boot partitions are created and the user must switch OSes from the bios boot menu. Do we want this to be the case, or should the two operating systems share a boot parititon?
- "Refresh" install failed from recovery partition. From an external drive the "Refresh" install worked successfully. 
- The password field presented when decrypting storage does not put focus on the password entry field (does not focus on any text box). 
- When looking at options available with Ubuntu already installed, when selecting "Alongside" or "Custom", the padding around the "Try Demo Mode" button changes (increases). Padding should be consistent.
- Installer logs for when refresh install failed of 18.10 on dualboot of 18.04 and 18.10: [installer.log](https://github.com/pop-os/installer/files/2637905/installer.log)


- I noticed on a refresh install on encrypted systems the location for possible refresh installs are listed similar to: `/dev/mapper/data-root` whereas if it is an unencrypted system, listed as: `/dev/sdx` or `/dev/nvmexny` I would think it might be more clear/easier to recognize an install from `/dev/sdx` rather than `/dev/mapper/data-root`. Is there anything that could be done here?

mmstick commented:
@brs17 

> On the "Resize OS" page, the left side of the installer ("Resize OS" text as well as Zoe image) is smaller than the left side content is on other pages.

Fixed by removing the elementary margins (48) from the content grid.

> The password field presented when decrypting storage does not put focus on the password entry field (does not focus on any text box).

Fixed the focus for the select and unlock buttons, as well as their sensitivities. Also added some constraints to the PV name entry so that users won't be able to input invalid characters.

> When looking at options available with Ubuntu already installed, when selecting "Alongside" or "Custom", the padding around the "Try Demo Mode" button changes (increases). Padding should be consistent.

Had the same issue as the partitioning view. That's fixed now.

> Just tried an "Alongside" install of Pop!_OS 18.04 with Pop!_OS 18.04 and got this failure: installer2.log

This only happens with 18.04? For whatever reason, fsck failed with "Operation not permitted".

> Installer logs for when refresh install failed of 18.10 on dualboot of 18.04 and 18.10: installer.log

Haven't been able to replicate so far. Will keep testing. I've added more context to the error messages in that region of distinst so that future logs will have a bit more detail on where this happened, exactly.

> The minimum size of the new install was 14.3GB (0 bytes free)

Do you think 5GB of overhead for a new install would be enough?

> I would think it might be more clear/easier to recognize an install from /dev/sdx rather than /dev/mapper/data-root. Is there anything that could be done here?

It's possible to resolve a LV to a VG, then to a PV, and then back to the original physical block device. I could add that behavior here.

> When successfully doing an "Alongside OS" install, two efi boot partitions are created and the user must switch OSes from the bios boot menu. Do we want this to be the case, or should the two operating systems share a boot parititon?

Supporting this just involves copying the contents of the other EFI partitions to the new EFI partition, if it is safe to do so. I did experiment with installing to an existing EFI partition. Would need to add a check to determine whether it's possible to re-use an existing EFI partition (free space + overhead).

We might also want to increase the default size of the EFI partition to 1G if we're supporting multiple operating systems + older kernels on the same EFI partition as Pop. Pop by itself uses about 50% of the 500M EFI partition, which leaves enough room for Windows (100M), but would not be enough for a dual boot of Pop or another Linux OS.

mmstick commented:
The help dialog should no longer open multiple instances now. GParted will also no longer open more than once if the modify button is clicked multiple times in advance.

mmstick commented:
> "Refresh" install failed from recovery partition

I suspect this is due to:

```rust
if let Some(ref recovery) = option.recovery_part.clone() {
    set_mount_by_identity(disks, recovery, "/recovery")?;
}
```
Recovery in this case being the active root mount. Fixing this now.

mmstick commented:
The refresh install from within a recovery partition is working. Currently fixing a few issues with performing an alongside install from recovery.

brs17 commented:
Although kinda confused in my comments, in 18.04 the Refresh logo is the same color as the mouse over hover and therefore not visible at times. When selected the logo is somewhat visible although not very.

brs17 commented:
Doing a refresh install from the recovery partition still fails due to `Device or resource busy` message.

brs17 commented:
After doing a fresh install of 18.04 and then rebooting into the live image to do an "Alongside" install, two "unused space on /dev/sda" appear. Once size is 18.4EB the other 120GB, /dev/sda is the live image media, so it should not be an option at all (let alone these incorrect sizes). The entire size of /dev/sda is 120GB so there isn't a partition that large either.

brs17 commented:
> 
> 
>     > The minimum size of the new install was 14.3GB (0 bytes free)
> 
> Do you think 5GB of overhead for a new install would be enough?

Perhaps we want to do a similar 20GB free? Although now we're talking about a not-insignificant amount of disk space.

mmstick commented:
- Found and fixed the last remaining corner case that could return an 18.4 EB partition.
- Distinst now creates logs beginning at the start of the application, rather than when the install starts.

Currently looking into the refresh on recovery issue.

mmstick commented:
I believe the refresh option now works within recovery. I'm going to test it again.

brs17 commented:
With "Install Alongside OS" it currently looks like there is some funkiness with resizing Windows, as well as an LVM partition and LUKS.

Per discussion, we think it makes sense to separate this pr into "Install alongside OS" and a separate PR for "Refresh install". "Refresh Install" is pretty close to being done and therefore can be introduced much sooner.
